scylla-cdc/src/checkpoints.rs

```diff
@@ -0,0 +1,192 @@
+use crate::stream_reader::StreamReader;
+use anyhow;
+use async_trait::async_trait;
+use futures::future::RemoteHandle;
+use futures::FutureExt;
+use scylla::frame::response::result::CqlValue::Timestamp;
+use scylla::prepared_statement::PreparedStatement;
+use scylla::Session;
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+use tokio::time::sleep;
+
+/// Starts creating checkpoints starting from now.
+/// Spawns a new task that periodically creates new checkpoint.
+/// Receives value of the next time window via channel.
+/// We are using tokio::sync::watch channel, because we only need the latest value.
+/// Returns handle to check on this newly created task.
+pub fn start_making_checkpoints(
+    mut checkpoint_maker: Box<dyn CDCCheckpointMaker>,
+    receiver: tokio::sync::watch::Receiver<Duration>,
+    saving_period: Duration,
+) -> RemoteHandle<anyhow::Result<()>> {
+    let (fut, handle) = async move {
+        loop {
+            let next_window_begin = *receiver.borrow();
+
+            // We are given timestamp of the beginning of the next read window,
+            // so we have to create a checkpoint with lower timestamp.
+            let checkpoint_timestamp =
+                Duration::checked_sub(next_window_begin, Duration::from_millis(1)).ok_or(
+                    anyhow::anyhow!(
+                        "Couldn't create a checkpoint, because starting timestamp was too low ( less than 1 ms )."
+                    ),
+                )?;
+            checkpoint_maker
+                .make_checkpoint(checkpoint_timestamp)
+                .await
+                .unwrap();
+            sleep(saving_period).await;
+        }
+    }
+    .remote_handle();
+
+    tokio::task::spawn(fut);
+
+    handle
+}
+
+/// Customizable trait responsible for creating checkpoints.
+#[async_trait]
+pub trait CDCCheckpointMaker: Send + Sync {
+    /// Creates new checkpoint with given timestamp.
+    async fn make_checkpoint(&mut self, timestamp: Duration) -> anyhow::Result<()>;
```

Comment:
Let's change this method to take `&self`. The user will probably provide a single instance of the maker and it will be shared by multiple consumers, and the user will be responsible for synchronization.

Alternatively, you can consider the factory pattern, just like CDCConsumerFactory + CDCConsumer. Then, you can keep the `&mut self`